File: Game/sangobuildtw/client/src/com/pip/io/UASegment.java
package com.pip.io;


import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import javax.microedition.io.Connector;
import javax.microedition.io.HttpConnection;

import com.pip.common.Tool;


public class UASegment{
    /** 协议类型 */
    public short type;
    /** 协议数据 */
    public byte[] data;
    public boolean handled;

    /** 协议串号标识，唯一 */
    public int serial = -1;
    public boolean needResponse;

    private int pos;
    private ByteArrayOutputStream segCache = null;
    private DataOutputStream outputCache = null;
    
    private static final Tool serialKey = new Tool();

    public UASegment(int type){
        this(type, false);
    }
    
    public UASegment(int type, boolean needSerial){
        this.type = (short)type;

        try{
            segCache = new ByteArrayOutputStream();
            outputCache = new DataOutputStream(segCache);
            outputCache.writeShort(type);
            
            if(needSerial) {
                serial = serialKey.nextKey();
                outputCache.writeInt(serial);        	
            }
        }catch(IOException ex){
        }
    }

    public UASegment(byte[] data){
        this.data = new byte[data.length];
        System.arraycopy(data, 0, this.data, 0, data.length);

        type = (short)getNumber(this.data, 0, 2);
        pos = 2;
    }

    public void reset(){
        pos = 2;
    }

    public void flush(){
        if(segCache == null){
            return;
        }

        try{
            outputCache.flush();
            data = segCache.toByteArray();
        }catch(Exception e){
        }finally{
            try{
                outputCache.close();
            }catch(Exception e){
            }
        }

        segCache = null;
        outputCache = null;

        setNumber(type, data, 0, 2);
    }

    public boolean readBoolean(){
        return (data[pos++] == 1)? true: false;
    }

    public byte readByte(){
        return data[pos++];
    }
    
    public int readUnsignedByte(){
        return data[pos++] & 0xFF;
    }

    public short readShort(){
        pos += 2;

        return (short)getNumber(data, pos - 2, 2);
    }
    
    public int readUnsignedShort(){
        return readShort() & 0xFFFF;
    }

    public int readInt(){
        pos += 4;

        return (int)getNumber(data, pos - 4, 4);
    }
    
    public void setInt(int num){
        setNumber(num, data, pos, 4);
    }

    public long readLong(){
        pos += 8;

        return getNumber(data, pos - 8, 8);
    }

    public String readString(){
        DataInputStream dis = null;
        try {
            if ((data[pos] & 0x80) == 0) {
                dis = new DataInputStream(new ByteArrayInputStream(data, pos, data.length - pos));
                pos += getNumber(data, pos, 2) + 2;
                return Tool.readUTF(dis);
            } else {
                int len = (int)(getNumber(data, pos, 2) & 0x7FFF);
                pos += 2;
                char[] charr = new char[len / 2];
                for (int i = 0; i < charr.length; i++) {
                    charr[i] = (char)(((data[pos] & 0xFF) << 8) | (data[pos + 1] & 0xFF));
                    pos += 2;
                }
                return new String(charr);
            }
        } catch (Exception e) {
            e.printStackTrace();

            return "";
        }finally{
            try{
                dis.close();
            }catch(Exception e){
            }
        }
    }

    public boolean[] readBooleans(){
        int len = (int)getNumber(data, pos, 2);
        pos += 2;
        boolean[] result = new boolean[len];

        for(int i = 0; i < len; i++){
            result[i] = readBoolean();
        }

        return result;
    }

    public byte[] readBytes(){
        int len = (int)getNumber(data, pos, 4);
        pos += 4;
        byte[] result = new byte[len];

        for(int i = 0; i < len; i++){
            result[i] = readByte();
        }

        return result;
    }

    public short[] readShorts(){
        int len = (int)getNumber(data, pos, 2);
        pos += 2;
        short[] result = new short[len];

        for(int i = 0; i < len; i++){
            result[i] = readShort();
        }

        return result;
    }

    public int[] readInts(){
        int len = (int)getNumber(data, pos, 2);
        pos += 2;
        int[] result = new int[len];

        for(int i = 0; i < len; i++){
            result[i] = readInt();
        }

        return result;
    }

    public long[] readLongs(){
        int len = (int)getNumber(data, pos, 2);
        pos += 2;
        long[] result = new long[len];

        for(int i = 0; i < len; i++){
            result[i] = readLong();
        }

        return result;
    }

    public String[] readStrings(){
        int len = (int)getNumber(data, pos, 2);
        pos += 2;
        String[] result = new String[len];

        for(int i = 0; i < len; i++){
            result[i] = readString();
        }

        return result;
    }

    public void writeBoolean(boolean b) throws IOException{
        outputCache.writeBoolean(b);
    }

    public void writeByte(byte b) throws IOException{
        outputCache.writeByte(b);
    }

    public void writeShort(short s) throws IOException{
        outputCache.writeShort(s);
    }

    public void writeInt(int n) throws IOException{
        outputCache.writeInt(n);
    }

    public void writeLong(long l) throws IOException{
        outputCache.writeLong(l);
    }

    public void writeString(String s) throws IOException{
        outputCache.writeUTF((s == null)? "": s);
    }

    public void writeBooleans(boolean[] b) throws IOException{
        outputCache.writeShort(b.length);

        for(int i = 0; i < b.length; i++){
            outputCache.writeBoolean(b[i]);
        }
    }

    public void writeBytes(byte[] b) throws IOException{
        outputCache.writeInt(b.length);

        for(int i = 0; i < b.length; i++){
            outputCache.writeByte(b[i]);
        }
    }

    public void writeShorts(short[] s) throws IOException{
        outputCache.writeShort(s.length);

        for(int i = 0; i < s.length; i++){
            outputCache.writeShort(s[i]);
        }
    }

    public void writeInts(int[] n) throws IOException{
        outputCache.writeShort(n.length);

        for(int i = 0; i < n.length; i++){
            outputCache.writeInt(n[i]);
        }
    }

    public void writeLongs(long[] l) throws IOException{
        outputCache.writeShort(l.length);

        for(int i = 0; i < l.length; i++){
            outputCache.writeLong(l[i]);
        }
    }

    public void writeStrings(String[] s) throws IOException{
        outputCache.writeShort(s.length);

        for(int i = 0; i < s.length; i++){
            outputCache.writeUTF(s[i]);
        }
    }

    public static long getNumber(byte[] buf, int off, int len){
        long longVal = 0;

        for(int i = 0; i < len; i++){
            longVal <<= 8;
            longVal |= buf[off + i] & 0xff;
        }

        return longVal;
    }

    public static void setNumber(int num, byte[] buf, int off, int len){
        for(int i = len - 1; i >= 0; i--){
            buf[off + i] = (byte)(num & 0xff);
            num >>= 8;
        }
    }

    /** 打开一个HTTP连接。 */
    public static HttpConnection getConnection(String url, boolean proxyFlag) throws IOException{
        String proxyUrl = null;

        if(proxyFlag){
            proxyUrl = "10.0.0.172:80";
        }else{
            proxyUrl = null;
        }

        // 应用代理
        String requestUrl = url, realHost = null;

        if(proxyUrl != null){
            int ind = url.indexOf('/', 7);

            if(ind >= 0){
                requestUrl = url.substring(0, 7) + proxyUrl + url.substring(ind);
                realHost = url.substring(7, ind);
            }
        }

        HttpConnection conn;
        //#if ModelID == Android || ModelID == Lenovo || ModelID == AndroidLarge || ModelID == AndroidSmall
		if(realHost != null) {
		    conn = (HttpConnection)Connector.open("p" + url);
		} else {
			conn = (HttpConnection)Connector.open(requestUrl);			
		}
        //#else
        // 打开连接        
        conn = (HttpConnection)Connector.open(requestUrl);

        if(realHost != null){
            conn.setRequestProperty("X-Online-Host", realHost);
        }
        //#endif

        return conn;
    }
}

File: web/client/src/vm/ua-segment.js
/*
 * The VM-side UASegment — com/pip/io/UASegment.java as the .etf scripts see it
 * through the 0x0071..0x008F syscall block.
 *
 * Unlike net/ua-codec.js (which splits read/write into two one-shot classes for
 * the wire layer), a script segment is a long-lived object: UWAP_Create makes it,
 * fields are written or read in place, and SendRequest ships it. So this class
 * keeps ONE growable buffer with a read cursor, exactly like the Java original
 * keeps outputCache + a byte[] to read back after flush().
 *
 * Layout: [ type:int16 ][ serial:int32 if needSerial ][ fields... ]
 */
import { encodeModifiedUTF, decodeModifiedUTF } from '../net/ua-codec.js';

let serialKey = 0;
export function nextSerialKey() {
  return ++serialKey;
}

export class UASegment {
  /**
   * @param {number} type opcode (int16)
   * @param {boolean} needSerial when true, an int serial is allocated and written
   */
  constructor(type, needSerial) {
    /** @type {number[]} */
    this.buf = [];
    this.type = type | 0;
    this.serial = -1;
    this.handled = false;
    this.needResponse = false;
    // reads start after the header THIS constructor wrote: type (+ serial).
    // Incoming segments built from received bytes keep pos = 2, where the
    // echoed serial (if any) is just the first field the script reads.
    this.startPos = needSerial ? 6 : 2;
    this.pos = this.startPos;
    this._writeShort(this.type);
    if (needSerial) {
      this.serial = nextSerialKey();
      this._writeInt(this.serial);
    }
  }

  /* ---- writes append at the end ------------------------------------- */

  _push(...bytes) {
    for (const b of bytes) this.buf.push(b & 0xff);
  }

  _writeShort(v) {
    this._push((v >> 8) & 0xff, v & 0xff);
  }

  _writeInt(v) {
    this._push((v >>> 24) & 0xff, (v >> 16) & 0xff, (v >> 8) & 0xff, v & 0xff);
  }

  writeBoolean(v) { this._push(v ? 1 : 0); }
  writeByte(v) { this._push(v); }
  writeShort(v) { this._writeShort(v); }
  writeInt(v) { this._writeInt(v); }

  /** length prefix first (that is what readString expects), then the body. */
  writeString(s) {
    const enc = encodeModifiedUTF(s == null ? '' : String(s));
    this._writeShort(enc.length - 2);
    for (let i = 2; i < enc.length; i++) this.buf.push(enc[i]);
  }

  /* ---- array writers: u16 count prefix, matching the read*() side ---- */

  writeBooleans(a) { this._writeShort(a.length); for (const x of a) this.writeBoolean(x); }
  writeBytes(a) { this._writeInt(a.length); for (let i = 0; i < a.length; i++) this.buf.push(a[i] & 0xff); }
  writeShorts(a) { this._writeShort(a.length); for (const x of a) this.writeShort(x); }
  writeInts(a) { this._writeShort(a.length); for (const x of a) this.writeInt(x); }
  writeStrings(a) { this._writeShort(a.length); for (const x of a) this.writeString(x); }

  /* ---- reads advance pos --------------------------------------------- */

  _need(n) {
    if (this.pos + n > this.buf.length) {
      throw new RangeError(`UASegment: read past end at ${this.pos} (+${n}) of ${this.buf.length}`);
    }
  }

  _readNum(len) {
    this._need(len);
    let v = 0;
    for (let i = 0; i < len; i++) v = (v << 8) | (this.buf[this.pos + i] & 0xff);
    this.pos += len;
    return v;
  }

  readBoolean() { this._need(1); return this.buf[this.pos++] === 1; }
  readByte() { this._need(1); const b = this.buf[this.pos++]; return b > 0x7f ? b - 0x100 : b; }
  readUnsignedByte() { return this._readNum(1); }
  readShort() { const v = this._readNum(2); return v > 0x7fff ? v - 0x10000 : v; }
  readUnsignedShort() { return this._readNum(2); }
  readInt() { return this._readNum(4) | 0; }

  /** UASegment.readString: modified-UTF8 branch OR high-bit UTF-16 branch. */
  readString() {
    this._need(2);
    if ((this.buf[this.pos] & 0x80) === 0) {
      const len = ((this.buf[this.pos] & 0xff) << 8) | (this.buf[this.pos + 1] & 0xff);
      this._need(2 + len);
      const bytes = Uint8Array.from(this.buf.slice(this.pos, this.pos + 2 + len));
      const { value } = decodeModifiedUTF(bytes, 0);
      this.pos += 2 + len;
      return value;
    }
    const len = this._readNum(2) & 0x7fff;
    this._need(len);
    let out = '';
    for (let i = 0; i < (len >> 1); i++) {
      out += String.fromCharCode(((this.buf[this.pos] & 0xff) << 8) | (this.buf[this.pos + 1] & 0xff));
      this.pos += 2;
    }
    return out;
  }

  readBooleans() { const n = this.readUnsignedShort(); const r = new Array(n); for (let i = 0; i < n; i++) r[i] = this.readBoolean(); return r; }
  readBytes() {
    const n = this.readInt();
    this._need(n);
    const r = Uint8Array.from(this.buf.slice(this.pos, this.pos + n));
    this.pos += n;
    return r;
  }
  readShorts() { const n = this.readUnsignedShort(); const r = new Array(n); for (let i = 0; i < n; i++) r[i] = this.readShort(); return r; }
  readInts() { const n = this.readUnsignedShort(); const r = new Array(n); for (let i = 0; i < n; i++) r[i] = this.readInt(); return r; }
  readStrings() { const n = this.readUnsignedShort(); const r = new Array(n); for (let i = 0; i < n; i++) r[i] = this.readString(); return r; }

  /**
   * Build the script-side view of an INCOMING segment — UASocketConnection.
   * readSegment's `new UASegment(buf)`: the whole wire segment (type header
   * included) is copied, type is read from the first two bytes, and the read
   * cursor starts at 2, where the echoed serial (if any) is just the first
   * field the script reads. `bytes` must already be a private copy — the
   * accumulator hands us one, and this constructor copies again so neither
   * side can observe the other's mutations.
   */
  static fromBytes(bytes) {
    const seg = new UASegment(0, false);
    seg.buf = Array.from(bytes, (b) => b & 0xff);
    seg.type = ((seg.buf[0] & 0xff) << 8) | (seg.buf[1] & 0xff);
    seg.startPos = 2;
    seg.pos = 2;
    return seg;
  }

  /* ---- lifecycle ------------------------------------------------------ */

  /** Java flush(): seal the written bytes so they can be read/sent. Here the
   *  buffer is always current, so this only rewinds the read cursor. */
  flush() { /* no-op */ }

  reset() { this.pos = this.startPos; }

  toBytes() {
    return Uint8Array.from(this.buf);
  }
}